商品購入機能のレビューをお願いいたします。

tsubi-kaiju opened the pull request:
# What
- 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
  - https://gyazo.com/4a70c686bfcf4f7dfc5a75ee68b2821d
- 入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
   - https://gyazo.com/4dcb27ebae7cd390c0156a2a522dc7b6
- ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
  - https://gyazo.com/233ca922d3c2deef119d805553d709ec
- ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
  - https://gyazo.com/0e1d164d7620c661e77c815bad6c9e99
- ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
  - https://gyazo.com/0ea8d7de940554409cd1f2781cf96dfc
- 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
  - https://gyazo.com/07eaa6c07a310dc04c58aca33970570c
- 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
  - https://gyazo.com/07eaa6c07a310dc04c58aca33970570c
- ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
  - 自分が出品した商品の場合
       - https://gyazo.com/07eaa6c07a310dc04c58aca33970570c
  - 他人が出品した商品の場合
       - https://gyazo.com/cf947020afe5bb86ced53e48eba279c0
- ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
  - https://gyazo.com/4ff350573ce81197f09055092cd892e1
- テスト結果の画像
  - https://gyazo.com/cf6b86ab84ae59176d2b00c151217c53
# Why
- 商品購入機能の実装のため

tsubi-kaiju committed:
Formオブジェクトの実装

Formオブジェクトを用いて、購入時にorderテーブルとaddressテーブルに情報を保存

tsubi-kaiju committed:
商品購入機能の関連部分を実装

【商品情報：Product】
・売却済みの商品は、画像上に「sold out」の文字が表示されるようになっていること（商品一覧機能）。
・売却済みの商品は、画像上に「sold out」の文字が表示されるようになっていること（商品詳細機能）。
・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されないこと（商品詳細機能）。
・ログイン状態の場合でも、自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること（商品情報編集機能）。

tsubi-kaiju committed:
PAY.JP(V2)を用いたクレジットカードでの決済機能

AsukaToyoshima commented:
レビューのご依頼ありがとうございます。
以下の動画につきまして、恐れ入りますが、再度撮影・添付していただけますでしょうか。
```
・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
→URLが購入ページにアクセスするものとなっておりました。編集ページのURLを入力お願いします。
```
上記を添付後、再度、レビュー依頼フォームの送信をお願いいたします。 お手数をおかけしますが、よろしくお願い致します。

tsubi-kaiju commented:
レビューありがとうございます。
エビデンスURLの誤り申し訳ございません。
以下の通り修正いたしましたので、ご確認お願いいたします。
> - ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）

  > - https://gyazo.com/4ff350573ce81197f09055092cd892e1

yuri240 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,78 @@
+        @order_address.telephone_number = ''
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Telephone number can't be blank")
+      end
```
telephone_numberについて以下のテストを追加で実行しましょう！
```
・9桁以下では登録できないこと
・12桁以上では登録できないこと
```

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ動画と修正点確認させていただきました。

コメントさせていただきましたので、ご確認お願いします。

また、お手数ではありますが
テストコードを追加した後は、改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。

tsubi-kaiju committed:
テスト項目を追加

以下のテストコードを追加
- telephone_numberが9桁以下では保存できないこと
- telephone_numberが12桁以上では保存できないこと

tsubi-kaiju commented:
ご指摘いただいたテスト項目を追加してテスト画像URLを更新しました。
ご確認お願いいたします。

> - テスト結果の画像
>   - https://gyazo.com/cf6b86ab84ae59176d2b00c151217c53

ushi10 reviewed (commented):
レビュー依頼ありがとうございます。確認いたしました。
こちら、LGTMとさせていただきます。実装お疲れさまでした。

改めて挙動をご確認いただき問題がない様子でしたらこちらのブランチを忘れずマージしましょう。
マージ後、ローカルのmasterブランチにて忘れずにプルをしましょう。 その後、新しいブランチを作成しissueを確認しながら実装に取り組みましょう。
